File: 155.cpp
#include <climits>
#include <cmath>
#include <iostream>
#include <stack>
#include <vector>


using  namespace  std;


class MinStack{
    public:

    stack<int> sta,minSta;




    MinStack(){
        
    }


    void push(int val){
        sta.push(val);

        if(minSta.empty() || minSta.top() >= val){
            minSta.push(val);
        }else {
            minSta.push(minSta.top());

        }



    }

    void pop(){
        sta.pop();
        minSta.pop();
    }

    int top()
    {
        return sta.top();
    }

    int getMin(){

        return minSta.top();
    }




};


int main()
{

MinStack minStack;
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
cout<<minStack.getMin()<<endl; // return -3
minStack.pop();
cout<<minStack.top()<<endl;    // return 0
cout<<minStack.getMin()<<endl; // return -2







    return 0;
}

File: 155.py
import heapq
import heapq

from typing import List


class MinStack:
    def __init__(self):
        self.sta = []
        self.staMinVal = []
        
    def push(self, val : int)-> None:
        self.sta.append(val)
        
        if not self.staMinVal:
            self.staMinVal.append(val)
        else:
            self.staMinVal.append(min(self.staMinVal[-1] , val))
            
    def pop(self) -> None:
        self.staMinVal.pop(-1)
        self.sta.pop(-1)
        
    
    
    def top(self) -> int:
        return self.sta[-1]
    
    def getMin(self) -> int:
        return self.staMinVal[-1]
    
    
    


MinS = MinStack()

MinS.push(-2)
MinS.push(0)
MinS.push(-3)
print(MinS.getMin()) # return -3
MinS.pop()
print(MinS.top())    # return 0
print(MinS.getMin()) # return -2

        
        